woocommerce/woocommerce
# Always sanitize coupon code to prevent inconsistent between admins and shop owners

### All Submissions:






### Changes proposed in this Pull Request:



By default WordPress will run `kses_init_filters()` if the current user can't use unfiltered HTML, applying the `wp_filter_kses()` function to the `title_save_pre`.
That's why admins can save a coupon code like `a&a` without getting converted into `a&amp;a`.

To fix the inconsistent I'm changing to always apply the same sanitation for all users on WooCommerce, also should be safer to always sanitize.
And it also includes a upgrade routine to fix all codes at once.

Closes #23655.

### How to test the changes in this Pull Request:

1. Create two coupons with the following codes using an admin:
- `a&a`
- `b&amp;b`
2. Now create more two coupons using a shop manager:
- `c&c`
- `d&amp;d`
3. Check that with the shop manager ampersand gets encoded.
4. With the shop manager edit the coupon with code `a&a` and note that the ampersand got encoded making impossible to restore.
5. Apply this patch, edit or create coupons with any user and see that ampersands always get encoded.
6. Added some products to the cart and try use your coupons created in the previous steps.
7. Also try apply coupons with both formats: `a&a` and `a&amp;a`.

### Other information:




### Changelog entry

> Fix - Coupon code inconsistent between admins and shop owners.
